waspc/waspls/src/Wasp/LSP/Completions/Common.hs

```diff
@@ -0,0 +1,52 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module Wasp.LSP.Completions.Common
+  ( CompletionProvider,
+    CompletionContext (..),
+    src,
+    cst,
+    makeBasicCompletionItem,
+  )
+where
+
+import Control.Lens (makeClassy)
+import qualified Data.Text as Text
+import qualified Language.LSP.Types as LSP
+import Wasp.Analyzer.Parser.CST (SyntaxNode)
+import Wasp.Analyzer.Parser.CST.Traverse (Traversal)
+
+-- | A function that providers 'LSP.CompletionItems' at a location.
+type CompletionProvider m = Traversal -> m [LSP.CompletionItem]
+
+-- | The context given to 'CompletionProvider's, via @MonadReader CompletionContext@.
```

Comment:
I think `MonadReader` is better if you have a lot of nested functions that need access to the same information, especially when some of the functions don't need it. That way you avoid having to pass the parameters down every level. Maybe a similar idea to using Contexts in React?

So I think making the decision comes down to: is it getting tedious to "thread" this argument through all these functions?

Reply:
I think that makes sense. Probably it would make sense for `Generator` in Wasp. That said, I really like seeing which arguments function receives, so I feel like situation has to get really ugly before I would go for doing this hah! But ok, if there are more parameters then it also makes sense (although they could always be grouped hm).

Reply:
I think `AppSpec` would be a decent candidate for putting inside a `MonadReader` in generator, could allow you to simplify stuff like:

```hs
genApp spec = genWebApp spec <++> genServer spec <++> genDb spec <++> genDockerFiles <++> genConfigFiles spec
```

into

```hs
genApp = genWebApp <++> genServer <++> genDb <++> genDockerFiles <++> genConfigFiles
```

but I agree it still doesn't really feel necessary here and also it feels a little magical. Potentially it would be good if you realize you actually need some more pieces of `AppSpec` in a function that you only passed parts of it to, without Reader you would have to go back and update all the function signatures with the new type that you want to pass down.